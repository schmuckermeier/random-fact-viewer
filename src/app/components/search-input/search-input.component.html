<form>
  <label>Search in favorites:
    <mat-form-field>
      <input [formControl]="formControl"
             [matAutocomplete]="auto"
             matInput
             placeholder="Start typing..."
             type="text"
             id="searchInput">
      <mat-autocomplete #auto="matAutocomplete" [displayWith]="displayFactText">
        @for (fact of filteredOptions | async; track fact.id) {
          <mat-option [value]="fact" (onSelectionChange)="onOptionSelected(fact)">{{ fact.text }}</mat-option>
        }
      </mat-autocomplete>
    </mat-form-field>
  </label>
</form>
