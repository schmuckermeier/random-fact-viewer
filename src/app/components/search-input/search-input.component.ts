import {Component, EventEmitter, inject, Output} from '@angular/core';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatInputModule} from '@angular/material/input';
import {AsyncPipe} from '@angular/common';
import {combineLatest, map, Observable, startWith} from 'rxjs';
import {FormControl, ReactiveFormsModule} from '@angular/forms';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {RandomFact} from '../../services/random-fact';

@Component({
  selector: 'app-search-input',
  imports: [MatFormFieldModule, MatAutocompleteModule, MatInputModule, AsyncPipe, ReactiveFormsModule],
  templateUrl: './search-input.component.html'
})
export class SearchInputComponent {
  formControl = new FormControl('');
  filteredOptions: Observable<RandomFact[]>;

  private favoriteFactsService = inject(FavoriteFactsService);

  @Output() selectedFact = new EventEmitter<RandomFact>();

  constructor() {
    const allFavoriteFacts = this.favoriteFactsService.getFacts();

    this.filteredOptions = combineLatest([
      allFavoriteFacts,
      this.formControl.valueChanges.pipe(startWith('')),
    ]).pipe(
      map(([allFacts, searchValue]) =>
        allFacts.filter(fact => {
            if(typeof searchValue !== 'string') {
              return []
            }
            return fact.text.toLowerCase().includes(searchValue ? searchValue.toLowerCase() : '')
          }
        )
      )
    );
  }

  onOptionSelected(fact: RandomFact) {
    this.selectedFact.emit(fact);
  }

  displayFactText(fact: RandomFact): string {
    return fact.text;
  }
}
