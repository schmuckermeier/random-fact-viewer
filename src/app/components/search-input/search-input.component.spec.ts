import { ComponentFixture, TestBed } from '@angular/core/testing';
import { SearchInputComponent } from './search-input.component';
import {of} from 'rxjs';
import {mockRandomFacts} from '../../testing/mockRandomFacts';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {queryById} from '../../testing/testing-utils';
import {By} from '@angular/platform-browser';
import {ReactiveFormsModule} from '@angular/forms';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatInputModule} from '@angular/material/input';
import {NO_ERRORS_SCHEMA} from '@angular/core';

describe('SearchInputComponent', () => {
  let component: SearchInputComponent;
  let fixture: ComponentFixture<SearchInputComponent>;

  let mockFavoriteFactsService = {
    getFacts: jest.fn(() => of(mockRandomFacts))
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        SearchInputComponent,
        ReactiveFormsModule,
        MatAutocompleteModule,
        MatFormFieldModule,
        MatInputModule
      ],
      providers: [
        {provide: FavoriteFactsService, useValue: mockFavoriteFactsService}
      ],
      schemas: [NO_ERRORS_SCHEMA]
    })
    .compileComponents();

    fixture = TestBed.createComponent(SearchInputComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should render correctly', () => {
    expect(fixture).toMatchSnapshot();
    expect(mockFavoriteFactsService.getFacts).toHaveBeenCalled();
  });

  it('should show all options', async () => {
    const input = queryById(fixture, "searchInput").nativeElement;
    input.focus();
    fixture.detectChanges();

    // Allow autocomplete dropdown to open
    await fixture.whenStable();
    fixture.detectChanges();

    const options = fixture.debugElement.queryAll(By.css('mat-option'));
    expect(options.length).toBe(3);
    expect(options[0].nativeElement.textContent).toContain(mockRandomFacts[0].text);
    expect(options[1].nativeElement.textContent).toContain(mockRandomFacts[1].text);
    expect(options[2].nativeElement.textContent).toContain(mockRandomFacts[2].text);
  });

  it('should select first options', async () => {
    const input = queryById(fixture, "searchInput").nativeElement;

    jest.spyOn(component.selectedFact, 'emit');

    input.focus();
    input.value = "Golf courses cover";
    input.dispatchEvent(new Event('input'));
    fixture.detectChanges();

    // Allow autocomplete dropdown to open
    await fixture.whenStable();
    fixture.detectChanges();

    const options = fixture.debugElement.queryAll(By.css('mat-option'));
    expect(options.length).toBe(1);
    const firstRandomFact = mockRandomFacts[0];
    expect(options[0].nativeElement.textContent).toContain(firstRandomFact.text);

    options[0].triggerEventHandler('onSelectionChange', { isUserInput: true });
    fixture.detectChanges();

    expect(component.selectedFact.emit).toHaveBeenCalledWith(firstRandomFact);
  });
});
