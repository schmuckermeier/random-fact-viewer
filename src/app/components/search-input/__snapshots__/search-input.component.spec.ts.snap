// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInputComponent should render correctly 1`] = `
<app-search-input
  favoriteFactsService={[Function Object]}
  filteredOptions={[Function Observable]}
  formControl={[Function FormControl2]}
  selectedFact={[Function EventEmitter_]}
>
  <form
    novalidate=""
  >
    <label>
      Search in favorites: 
      <mat-form-field
        class="mat-mdc-form-field mat-mdc-form-field-type-mat-input mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-valid"
      >
        <div
          class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled mdc-text-field--no-label"
        >
          <div
            class="mat-mdc-form-field-focus-overlay"
          />
          <div
            class="mat-mdc-form-field-flex"
          >
            <div
              class="mat-mdc-form-field-infix"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-invalid="false"
                aria-required="false"
                autocomplete="off"
                class="mat-mdc-autocomplete-trigger mat-mdc-input-element mat-mdc-form-field-input-control mdc-text-field__input ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                id="searchInput"
                matinput=""
                placeholder="Start typing..."
                role="combobox"
                type="text"
              />
              <mat-autocomplete
                class="mat-mdc-autocomplete"
              >
              </mat-autocomplete>
            </div>
          </div>
          <div
            class="mdc-line-ripple mdc-line-ripple--deactivating"
            matformfieldlineripple=""
          />
        </div>
        <div
          class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="mat-mdc-form-field-hint-wrapper"
          >
            <div
              class="mat-mdc-form-field-hint-spacer"
            />
          </div>
        </div>
      </mat-form-field>
    </label>
  </form>
</app-search-input>
`;
